### PR TITLE
Add criticality and austerity system

### PR DESCRIPTION
**Pull request by ChimeraCoder**
#### Summary

Add the criticality and austerity system to the open-source Unilog project.

This is slightly a WIP. I think I'm going to move the clevel definitions to a separate package, so that we can import them elsewhere easily.

I also need to figure out the vendoring issue (see below). Otherwise, though, this is complete and tested (locally).

Before merging, I need to rebase to get rid of the intermediate commits and add Kiran as a contributor since she worked on the sampling portion of the original code.

Vendoring:

At the moment, this builds on our CI (as does the branch of `unilog.stripe` which imports this, after updating the vendored libraries), which is what we ultimately need.

However, `unilog.stripe` doesn't build locally, because it tries to import the vendored libraries in *this* repository, and then it runs into a type error with the two "different" versions of the statsd client library. The only way to solve it is to delete the `vendor` directory in this repository (locally) and then build the `unilog.stripe` project.

I'd like to figure out the right incantation for solving this general problem, because I'm pretty sure that one exists. However, barring that, we could just use a local interface type to fix the issue.

#### Test plan

Added tests.

Tested the sample binary locally using bash function to dump text to console.

#### Rollout/monitoring/revert plan

r? @kiran-stripe (co-author, pairing)
cc @stripe/observability

### CODE REVIEW

**Comment by ChimeraCoder** [comment]
thanks gerald

**Comment by ChimeraCoder** [comment]
okay, this is ready now!

File: unilog/generate_input.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  echo "this is a default (sheddableplus)"
+}
+
+
+
+while true; do
+  dumpText
+  sleep 2
+done
+
```
Comment (kiran-stripe) [inline unilog/generate_input.sh]: did you mean to commit this test file?

File: clevels/austerity_test.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
...west-1.apiori.com/EzBDuA4iNq-2631925524 85137cc252d87354>e9b8c49860f01f15] CANONICAL-API-LINE: api_method=AccountRetrieveMethod content_type="application/x-www-form-urlencoded" created=1478805073.5253563 http_method=GET ip="54.210.12.235" path="/v1/accounts/acct_xxxxxxxxxxxxxxxx" user_agent="Stripe/v1 RubyBindings/1.31.0" request_id=req_xxxxxxxxxxxxxx response_stripe_version="2016-03-07" status=200 merchant=acct_xxxxxxxxxxxxx merchant__created=1474909410.2272582 merchant__stripe_version="2016-07-06" logical_shard=maindb_0030 physical_shard=shard_k key_id=mk_xxxxxxxxxxxxx livemode=true perms=admin perms_used="limited_account_read,account_admin_read" application=ca_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx page_mode=none idempotency_work_done=true limiting_considered=true describe_duration=0.175 duration=0.234 end_to_end_duration=0.236 lock_duration=0.002 wabbit_duration=0.001 write_event_duration=0.002 publish_kafka_duration=0.009 cpu_duration=0.066 bindings_language=ruby bindings_language_version="2.2.3 p173 (2015-08-18)" bindings_version="1.31.0" git_revision=5e00cf28 tls_cipher="ECDHE-RSA-AES128-GCM-SHA256" tls_version="TLSv1.2" alloc_count=54187 gc_duration=0.005 duration_publish=0.001`,
```
Comment (kiran-stripe) [inline clevels/austerity_test.go]: maybe a liiittle more redaction/a narrower test case here

**Comment by kiran-stripe** [comment]
lgtm, pending 1-2 nits!

**Comment by ChimeraCoder** [comment]
@kiran-stripe ready for re-r!


If this is good, let me know (but don't mark as approved yet, so we don't merge by mistake). I'll update the contributor information so it gets both of us once it's done, and then self-approve before merging.

**Comment by kiran-stripe** [comment]
lgtm

File: unilog/generate_input.sh (hunk 1 above)
Comment (ChimeraCoder) [inline unilog/generate_input.sh]: Yes! This is intended to generate input logged at different levels, to make it easy to test behavior locally.

**Comment by ChimeraCoder** [comment]
Self-approving after rebase, plus update to committer information for dc91466. (The portions being ported from the internal version are covered by 13f7fa0 and dc91466; the latter commit corresponds roughly to the exponential random sampling logic, which Kiran wrote, as contrasted with the fixed-rate sampling that we used in the previous version).

**Pull request merged**